updater/bootstrap.ts

```diff
@@ -0,0 +1,352 @@
+/**
+ * Updater bootstrap:
+ * Determines if the user should receive an update, and if so,
+ * spawns the updater and shuts down the application. In most
+ * error situations, the application will start up normally.
+ */
+
+import * as util from 'util';
+import * as path from 'path';
+import * as tasklist from 'tasklist';
+import * as fs from 'fs';
+import * as request from 'request';
+import * as cp from 'child_process';
+import * as semver from 'semver';
+import * as log from 'electron-log';
+import * as crypto from 'crypto';
+
+const readFile = util.promisify(fs.readFile);
+const writeFile = util.promisify(fs.writeFile);
+const mkdir = util.promisify(fs.mkdir);
+
+/**
+ * Use our own promisify for request, has better typings
+ */
+const prequest = (info: request.UriOptions & request.CoreOptions) => {
+  return new Promise<request.Response>((resolve, reject) => {
+    request(info, (err, response) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(response);
+      }
+    });
+  });
+};
+
+/**
+ * The main process provides this to the bootstrap module. It describes
+ * the state of the current version, as well as how, where from, and to
+ * where we should update.
+ */
+interface IUpdateInfo {
+  baseUrl: string;
+  version: string;
+  exec: string[];
+  cwd: string;
+  waitPids: number[];
+  appDir: string;
+  tempDir: string;
+  cacheDir: string;
+  versionFileName: string;
+}
+
+/**
+ * This describes the shape of a channel file (i.e. latest.json) living
+ * on our update CDN.
+ */
+interface ILatestVersionInfo {
+  /**
+   * The latest available version
+   */
+  version: string;
+
+  /**
+   * A random string chosen at release time that is added to the
+   * update id before hashing into a release bucket to ensure we
+   * rollout to a different sample of people with each release.
+   */
+  seed: string;
+
+  /**
+   * Determines the percentage of users this release is rolled out to
+   */
+  rollout: {
+    /**
+     * The default rollout percentage used when deciding whether to update
+     */
+    default: number;
+
+    /**
+     * A version-specific override that allows us to target this release at
+     * people running a specific version.
+     */
+    [version: string]: number;
+  };
+}
+
+/**
+ * Generates a relatively small and uniform numeric hash of a string
+ *
+ * Inspired by MIT licensed: https://github.com/segment-boneyard/hash-mod
+ * @param str The string to hash
+ */
+function hashToInt(str: string) {
+  return str.split('').reduce((hash, char) => {
+    return (hash * 31 * char.charCodeAt(0)) % 982451653;
+  }, 7);
+}
```

Comment:
Wow, that kind of dark magic

Reply:
Lol hashing algorithms are always dark magic + prime numbers haha.